### PR TITLE
add type hints

### PR DESCRIPTION
**Pull request by ambrosejcarr**
- Fixes some minor problems with my previous PRs related to typing
- Adds hints for uninitialized argparse.ArgumentParser objects to PipelineComponents
- reassign `id` -> `id_` to avoid clobbering builtin `id` method
- Assigned an errant TODO

### CODE REVIEW

File: starfish/pipeline/features/spots/detector/__init__.py [hunk 1]
```diff
@@ -35,7 +33,7 @@ def add_to_parser(cls, subparsers):
         cls.spot_finder_group = spot_finder_group
 
     @classmethod
-    def _cli(cls, args: argparse.Namespace, print_help=False):
+    def _cli(cls, args, print_help=False):
```
Comment (ttung) [inline starfish/pipeline/features/spots/detector/__init__.py]: why this change?

File: starfish/pipeline/pipelinecomponent.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import collections
+import argparse
```
Comment (ttung) [inline starfish/pipeline/pipelinecomponent.py]: sort order.

File: starfish/pipeline/pipelinecomponent.py [hunk 3]
```diff
@@ -7,6 +8,10 @@ class PipelineComponentType(type):
     This is the metaclass for PipelineComponent.  As each subclass that is _not_ PipelineComponent is created, it sets
     up a map between the algorithm name and the class that implements it.
     """
+
+    # this seems to be expected by this class as well.
+    _algorithm_to_class_map: Optional[Mapping[str, Type]] = None
```
Comment (ttung) [inline starfish/pipeline/pipelinecomponent.py]: that's pycharm being confused.  it shouldn't be required.

File: starfish/pipeline/features/spots/detector/__init__.py (hunk 1 above)
Comment (ambrosejcarr) [reply starfish/pipeline/features/spots/detector/__init__.py]: Makes pycharm super upset.

File: starfish/pipeline/pipelinecomponent.py (hunk 2 above)
Comment (ambrosejcarr) [reply starfish/pipeline/pipelinecomponent.py]: Ugh I thought I got them all.

File: starfish/pipeline/pipelinecomponent.py (hunk 3 above)
Comment (ambrosejcarr) [reply starfish/pipeline/pipelinecomponent.py]: Got it. I'll take that out.

File: starfish/pipeline/features/spots/detector/__init__.py (hunk 1 above)
Comment (ambrosejcarr) [reply starfish/pipeline/features/spots/detector/__init__.py]: If I could disable inspections on that parameter in the IDE, would that be a good middle ground?

Comment (ttung) [reply starfish/pipeline/features/spots/detector/__init__.py]: That's cool, just drop the type hint then.

It's frustrating that the type hinting system isn't better. :(

Comment (ambrosejcarr) [reply starfish/pipeline/features/spots/detector/__init__.py]: Agree. :(

File: starfish/pipeline/pipelinecomponent.py (hunk 3 above)
Comment (ttung) [reply starfish/pipeline/pipelinecomponent.py]: Oh, you need to rebase, I added this already.

**Comment by ambrosejcarr** [comment]
Ok, I've figured things out. Calling the tests generated by this PR produces the following error in the following case: 

This code block is added
```
class PipelineComponent(metaclass=PipelineComponentType):

    ...

    @classmethod
    def _cli(cls, args: argparse.Namespace):
        raise NotImplementedError()
```
without also adding:
```
```


```
$ make test
if [ "1" == 1 ]; then \
                STARFISH_COVERAGE=1 coverage run -p --source=starfish -m unittest tests.test_apply; \
        else \
                python -m unittest tests.test_apply; \
        fi
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/__main__.py", line 18, in <module>
    main(module=None)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/main.py", line 94, in __init__
    self.parseArgs(argv)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/main.py", line 141, in parseArgs
    self.createTests()
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/main.py", line 148, in createTests
    self.module)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/loader.py", line 219, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/loader.py", line 219, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/loader.py", line 153, in loadTestsFromName
    module = __import__(module_name)
  File "/Users/ambrosecarr/projects/spacetx/starfish/tests/test_apply.py", line 6, in <module>
    from starfish.constants import Coordinates, Indices
  File "/Users/ambrosecarr/projects/spacetx/starfish/starfish/__init__.py", line 1, in <module>
    from .starfish import starfish
  File "/Users/ambrosecarr/projects/spacetx/starfish/starfish/starfish.py", line 9, in <module>
    from .pipeline.features.spots.detector import SpotFinder
  File "/Users/ambrosecarr/projects/spacetx/starfish/starfish/pipeline/features/spots/detector/__init__.py", line 11, in <module>
    class SpotFinder(PipelineComponent):
  File "/Users/ambrosecarr/projects/spacetx/starfish/starfish/pipeline/pipelinecomponent.py", line 18, in __init__
    PipelineComponentType._ensure_algorithms_setup(cls)
  File "/Users/ambrosecarr/projects/spacetx/starfish/starfish/pipeline/pipelinecomponent.py", line 22, in _ensure_algorithms_setup
    if cls._algorithm_to_class_map is None:
AttributeError: type object 'SpotFinder' has no attribute '_algorithm_to_class_map'
make: *** [tests/test_apply.py] Error 1
```

File: starfish/pipeline/pipelinecomponent.py [hunk 4]
```diff
@@ -7,6 +8,9 @@ class PipelineComponentType(type):
     This is the metaclass for PipelineComponent.  As each subclass that is _not_ PipelineComponent is created, it sets
     up a map between the algorithm name and the class that implements it.
     """
+
+    _algorithm_to_class_map: Optional[Mapping[str, Type]] = None
```
Comment (ttung) [inline starfish/pipeline/pipelinecomponent.py]: this triggers an error for me.  removing it makes mypy happy.

Comment (ambrosejcarr) [reply starfish/pipeline/pipelinecomponent.py]: for me, removing it makes the _python_ tests die. I fixed it with `Optional[]` which passes both.

File: starfish/pipeline/features/_validated_table.py [hunk 5]
```diff
@@ -2,6 +2,9 @@
 
 
 class ValidatedTable:
+
+    required_fields = NotImplemented
```
Comment (ttung) [inline starfish/pipeline/features/_validated_table.py]: I don't think this is how NotImplemented is supposed to be used as per the docs but it's not a dealbreaker.

Comment (ambrosejcarr) [reply starfish/pipeline/features/_validated_table.py]: That's actually an interesting one -- I needed to define that for mypy to let me define load in the base class, since it requires that object. should I... make it a `property` and have the property return `NotImplementedError`? I don't think that would harm the inheritance pattern.

Comment (ttung) [reply starfish/pipeline/features/_validated_table.py]: I don't have a great solution for this.  I just slapped an annotation to bypass the check in my PR.

Comment (ttung) [reply starfish/pipeline/features/_validated_table.py]: https://github.com/spacetx/starfish/pull/227/files#diff-9bdf66c5007c8d6623fc4e46338b6b0aR67

Comment (ambrosejcarr) [reply starfish/pipeline/features/_validated_table.py]: I tried making it a property (see #230) but mypy got angry with me for reasons related to the property.
